--- a/lib/src/udt_port/list/loss_list.dart
+++ b/lib/src/udt_port/list/loss_list.dart
@@ -80,7 +80,8 @@

   void _coalesce() {
     var i = 0;
-    while (i < _intervals.length - 1) {
+    var len = _intervals.length;
+    while (i < len - 1) {
       final current = _intervals[i];
       final next = _intervals[i + 1];
       final adjacent = UdtSequenceNumber.increment(current.end) == next.start;
@@ -90,6 +91,7 @@
           current.end = next.end;
         }
         _intervals.removeAt(i + 1);
+        len--;
       } else {
         i++;
       }
@@ -204,7 +204,8 @@

   void _coalesce() {
     var i = 0;
-    while (i < _intervals.length - 1) {
+    var len = _intervals.length;
+    while (i < len - 1) {
       final current = _intervals[i];
       final next = _intervals[i + 1];
       final adjacent = UdtSequenceNumber.increment(current.end) == next.start;
@@ -214,6 +215,7 @@
           current.end = next.end;
         }
         _intervals.removeAt(i + 1);
+        len--;
       } else {
         i++;
       }
